### ISSUE TITLE
Add a method akin to cursor.hasNext()

### ISSUE THREAD

**Issue by andyburke** [rethinkdb/rethinkdb#4517]
It would be nice to have a simple, straightforward way to know if a cursor has any (more) items. 

Coercing it to an array and checking for that array to be empty is not (I wouldn't think) efficient, especially if there many records.

The existing each() and next() methods don't make this easy or straightforward.

The use case I am looking at right now is checking for existing users in a database who have the same email address or phone number. Currently, I am doing:

```javascript
        var error = null;
        var existing = null;
        cursor.each( function( _error, _user ) {
            error = _error;
            existing = _user;
            cursor.close();
        }, function() {
            if ( error ) {
                callback( error );
                return;
            }

            if ( existing ) {
                callback( 'Existing user!' );
            }

            callback();
        } );
```

It would be much nicer to be able to do:

```javascript
    if ( cursor.hasNext() ) {
        callback( 'Existing user!' );
        return;
    }
```

If, however, it needs to be asynchronous, it still seems more straightforward than the each()/cursor.close() construction above:

```javascript
    cursor.hasNext( function( error, hasNext ) {
        callback( hasNext ? 'Existing user!' : error );
    } );
```

**Comment by thejsj** [comment]
I agree that this would be useful. I got [this question on StackOverflow](http://stackoverflow.com/questions/29585123/rethinkdb-determine-if-there-is-a-next-document-in-result-set) and the answer is a bit cumbersome.

**Comment by neumino** [comment]
You can just retrieve the user or an error:
```
query.nth(0).default(r.error("User not found")).run()
```

**Comment by andyburke** [comment]
That doesn't really feel right, either...

**Comment by neumino** [comment]
If the error bothers you, you can just return `null`:
```
query.nth(0).default(null)
```

Also this
```
    cursor.hasNext( function( error, hasNext ) {
        callback( hasNext ? 'Existing user!' : error );
    } );
```
Is the same (if you return documents) as
```
    cursor.next( function( error, user ) {
        callback( !!user ? 'Existing user!' : error );
    } );
```

**Comment by andyburke** [comment]
These feel better, but I still wonder if a hasNext() would be a more semantic alternative.

**Comment by mlucy** [comment]
One of the problems with `hasNext()` is that it sometimes requires blocking on data from the server, so it would have to take a callback in JS, which would be a little awkward.

**Comment by andyburke** [comment]
This:

```javascript
cursor.next( function( error, user ) {
    callback( !!user ? 'Existing user!' : error );
} );
```

Gives me this, even with a .default( null ) on the query:

```
No more rows in the cursor.
```

What am I missing? This just seems so much more complex than it should be.

**Comment by danielmewes** [comment]
That just seems to be how the callback is written. If there is no existing user, it calls the callback with the error. You could check what type the error has in order to distinguish query errors from the case where there are no results.

I think `query.nth(0).default(null)` is really the best option though.

**Comment by coffeemug** [comment]
To expand on @mlucy's answer, we discussed this before and this turns out to be a surprisingly tricky issue.

It's not possible to add simple `hasNext` and `isEmpty` methods to the cursor, because the client doesn't always know if there is more data to be fetched from the database server. So these functions would need to take a callback and return results asynchronously, which means two things: you couldn't write a simple loop with `hasNext` as a check (you'd have to deal with a callback), and since each call to `hasNext` would have to go to the server, the performance of a naive loop you might expect the user to write would be suboptimal.

So we decided to keep these methods out, and instead give different guidance on how to use cursors. It is a little bit inconvenient at the beginning (sorry about that!), but unfortunately it's the best way to do it because of the above technical limitations.

Hope this helps @andyburke! (We'd love to send you a t-shirt to compensate if you fill out http://rethinkdb.com/community/shirts-for-stories/ :smile:)

**Comment by andyburke** [comment]
I understand that hasNext() would need to be async. That doesn't bother me that much.

I realize this would be a compatibility issue, but another alternative I think would be more obvious would be to make next() not call back with an error when the cursor is empty. I mean, cursors usually eventually run out of things. That doesn't seem like an exceptional (error) case to me. The only way I can think of that it would be a problem is if you expected null documents, which seems unlikely. At least then I could do:

```javascript
cursor.next( function( error, user ) {
    callback( !!user ? 'Existing user!' : error );
} );
```

The alternative of checking for a particular error seems wrong.

Also, in the case where i am calling .next() do I need to call cursor.close()?

**Comment by coffeemug** [comment]
@andyburke -- it helps to call `cursor.close()` (though if you don't, nothing catastrophic should happen).

We could add an optarg to `next` that tells it not to call the error callback on finish (and potentially make it the default). Does anyone have an opinion on that? We want to be compatible with other Node.js database APIs, though, and I think the way we do it now is usually a standard API.

**Comment by mlucy** [comment]
I would't be opposed to offering an optarg, but I'm not sure about making it the default.  I would personally expect `next` to tell me if there was no next element rather than silently ignoring the error.

**Comment by andyburke** [comment]
@mlucy isn't getting a null element back the indication it is out of elements?

I guess I'm saying: errors should be exceptional cases, not expected cases. cursors are expected to run out of elements (almost always).

**Comment by mlucy** [comment]
@andyburke -- it's normal to get back `null` elements.  For example if you have a table `test` with the three documents `{field: 'foo'}`, `{field: null}`, `{field: 'bar'}` and you write `r.table('test')('field')` you'll get back the stream `'foo', null, 'bar'`.

It's true that you expect to reach the end of the cursor at some point if you're just calling `next` over and over again.  But if you're planning to iterate over the whole stream until you reach the end, `each` is probably the right tool for that.

**Comment by danielmewes** [comment]
Marking as a ReQL_proposal for an optarg for `next`.

Would it actually make it any easier to handle though if `next` called the callback with `null` and no error rather than `null` and an error?

**Comment by andyburke** [comment]
Would an undefined value work better than null?

If you can always expect any value to come back from the cursor and for that to be potentially valid, then maybe there is no better way than having an error. This just feels really wrong to me:

```javascript
```

**Comment by andyburke** [comment]
Maybe it's better to loop back to this proposal:

```javascript
query.nth(0).default(null)
```

That just isn't intuitive to me, but something like this would be reasonable IMO:

```javascript
query.exists( function( error, exists ) {
  // error if there's a db problem of some kind
  // exists == true if the query matched at least one document, false otherwise
} );
```

Internally, exists could just be the query.nth(0).default(null) implementation mentioned above, but a .exists() feels way more intuitive for end users to me.

**Comment by marshall007** [comment]
@andyburke if you don't need the document back you can just do `query.isEmpty()`.
